src/V3Increments.cpp

```diff
@@ -0,0 +1,204 @@
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Global.h"
+#include "V3Increments.h"
+#include "V3Ast.h"
+
+#include <algorithm>
+#include <cstdarg>
+#include <vector>
+
+#include <iostream>
+
+//######################################################################
+
+class IncrementsVisitor : public AstNVisitor {
+private:
+    // STATE
+    int                 m_modIncrementsNum;  // Var name counter
+
+    // METHODS
+    void insertBefore(AstNode* placep, AstNode* newp) {
+        AstNRelinker linker;
+        placep->unlinkFrBack(&linker);
+        newp->addNext(placep);
+        linker.relink(newp);
+    }
+
+public:
+    // CONSTRUCTORS
+    explicit IncrementsVisitor(AstNetlist* nodep) {
+        m_modIncrementsNum = 0;
+        iterate(nodep);
+    }
+
+    virtual ~IncrementsVisitor() {}
+
+    virtual void visit(AstPreInc* nodep) VL_OVERRIDE {
+        iterateChildren(nodep);
+
+        AstNodeVarRef *vr = VN_CAST(nodep->op1p(), VarRef);
+        AstNode *back = nodep->backp();
+
+        nodep->replaceWith(new AstVarRef(
+                    back->fileline(),
+                    vr->varp(),
+                    false));
+
+        VL_DO_DANGLING(nodep->deleteTree(), nodep);
+
+        AstNode *replace = back->op1p();
+        FileLine *fl = replace->fileline();
+
+        insertBefore(back->backp(),
+            new AstAssign(fl,
+                new AstVarRef(fl, vr->varp(), true),
+                new AstAdd(fl,
+                    new AstVarRef(fl, vr->varp(), false),
+                    new AstConst(fl, 1)
+                )
+            )
+        );
+    }
+
+    virtual void visit(AstPreDec* nodep) VL_OVERRIDE {
+        iterateChildren(nodep);
+
+        AstNodeVarRef *vr = VN_CAST(nodep->op1p(), VarRef);
+        AstNode *back = nodep->backp();
+
+        nodep->replaceWith(new AstVarRef(
+                    back->fileline(),
+                    vr->varp(),
+                    false));
+
+        VL_DO_DANGLING(nodep->deleteTree(), nodep);
+
+        AstNode *replace = back->op1p();
+        FileLine *fl = replace->fileline();
+
+        insertBefore(back->backp(),
+            new AstAssign(fl,
+                new AstVarRef(fl, vr->varp(), true),
+                new AstSub(fl,
+                    new AstVarRef(fl, vr->varp(), false),
+                    new AstConst(fl, 1)
+                )
+            )
+        );
+    }
+
+    virtual void visit(AstPostInc* nodep) VL_OVERRIDE {
+        iterateChildren(nodep);
+
+        AstNodeVarRef *vr = VN_CAST(nodep->op1p(), VarRef);
+        AstNode *back = nodep->backp();
+
+        AstNode *replace = back->op1p();
+        FileLine *fl = replace->fileline();
+
+        string name = string("__Vincrement")+cvtToStr(m_modIncrementsNum++);
+
+        AstVar* varp = new AstVar(fl, AstVarType::BLOCKTEMP, name,
+                                  back->backp()->findSigned32DType());
```

Comment:
Need to use the data type of the original signal, nodep->dtypep()

Reply:
I'm getting NULL from dtypep(), I've tried using it on nodep, as well as other nodes (Var, VarRef) and for all of them it's returning NULL, is it possible that dtypes are not set at this stage of process function?

Reply:
Ah, right, so in this case call ->subDTypep() which will get it from under the old variable (where it lives until it gets to V3Width, then moves to dtypep()).